contracts/contracts/Umbra.sol

```diff
@@ -198,11 +209,82 @@ contract Umbra is Ownable {
     bytes32 _r,
     bytes32 _s
   ) external {
+    _validateWithdrawSignature(_stealthAddr, _acceptor, _tokenAddr, _sponsor, _sponsorFee, UmbraHookable(0), "", _v, _r, _s);
+    _withdrawTokenInternal(_stealthAddr, _acceptor, _tokenAddr, _sponsor, _sponsorFee, UmbraHookable(0), "");
+  }
+
+  /**
+   * @notice Withdraw an ERC20 token payment on behalf of a stealth address via signed authorization
+   * @param _stealthAddr The stealth address whose token balance will be withdrawn
+   * @param _acceptor Address where withdrawn funds should be sent
+   * @param _tokenAddr Address of the ERC20 token being withdrawn
+   * @param _sponsor Address which is compensated for submitting the withdrawal tx
+   * @param _sponsorFee Amount of the token to pay to the sponsor
+   * @param _hook Contract that will be called after the token withdrawal has completed
+   * @param _data Arbitrary data that will be passed to the post-withdraw hook contract
+   * @param _v ECDSA signature component: Parity of the `y` coordinate of point `R`
+   * @param _r ECDSA signature component: x-coordinate of `R`
+   * @param _s ECDSA signature component: `s` value of the signature
+   */
+  function withdrawTokenAndCallOnBehalf(
+    address _stealthAddr,
+    address _acceptor,
+    address _tokenAddr,
+    address _sponsor,
+    uint256 _sponsorFee,
+    UmbraHookable _hook,
+    bytes memory _data,
+    uint8 _v,
+    bytes32 _r,
+    bytes32 _s
+  ) external {
+    _validateWithdrawSignature(_stealthAddr, _acceptor, _tokenAddr, _sponsor, _sponsorFee, _hook, _data, _v, _r, _s);
+    _withdrawTokenInternal(_stealthAddr, _acceptor, _tokenAddr, _sponsor, _sponsorFee, _hook, _data);
+  }
+
+  /// @dev low level withdrawal function that should only be called after safety checks
+  function _withdrawTokenInternal(
+    address _stealthAddr,
+    address _acceptor,
+    address _tokenAddr,
+    address _sponsor,
+    uint256 _sponsorFee,
+    UmbraHookable _hook,
+    bytes memory _data
+  ) internal {
     uint256 _amount = tokenPayments[_stealthAddr][_tokenAddr];
 
-    // also protects from underflow
+    // also protects from underflow and re-entrance
```

Comment:
Hmm, yep, I think you're right. Since we no longer explicitly prevent stealth address re-use, they could re-enter by first sending more tokens to the same address. Shucks, I thought this was a useful side effect.

I also think you're right about the impact of re-entrance here, but this is probably something we want to flag explicitly for the auditors.

Reply:
Did we ever create an issue to track notes for auditors? We had a few other things we wanted to mention, though I can't recall what they were at the moment 😬

Reply:
Now tracked in #123